사다리게임 3단계 - 객체 역할 분담

popsmile opened the pull request.

popsmile committed:
연결된 행 검사및삭제 과정에 대한 여러 방법들 중 하나 선택하여 나머지 삭제

popsmile committed:
필요한 구조체 생성 후, 개별 파일로 분리

popsmile committed:
LadderGame 변수명 변경 및 InputView 완성

popsmile committed:
InputView, LadderStep 필요없는 주석 제거

popsmile committed:
LadderGame의 사다리배열 생성부분 및 ResultView의 출력부분 작성

popsmile committed:
main에 resultView적용

godrm commented on LadderGame/LadderGame/InputView.swift:
```diff
@@ -0,0 +1,64 @@
+
+struct InputView {
+    var guideMessage : [String]
+    var ladderGame : LadderGame
```
- struct 를 사용할 때는 항상 private 를 먼저 고려하세요.
구조체 외부에서 내부 속성에 그대로 접근해서 값을 바꿀 수 있으면 부작용을 일으킬 수 있습니다. 
- LadderGame 인스턴스를 InputView가 포함해야 할까요? 
특히 핵심 로직을 구현한 객체와 IntputView나 OutputView와 이런 포함관계를 분리하는 게 좋습니다. LadderGame은 상위 모듈로 옮겨서 InputView와 대등하게 수준으로 만드세요.

godrm commented on LadderGame/LadderGame/InputView.swift:
```diff
@@ -0,0 +1,64 @@
+    }
+    
+    // 위 문자열과 문자열배열로 ladderGame 값을 변경해주는 함수
+    mutating func writeInfoOf(_ maxHeight:Int, _ playerNames:[String]) {
```
InputView 메소드들 중에서도 외부에서 사용하지 않고 내부에서만 사용하는 경우는 private 처리를 꼭 해주세요.

godrm commented on LadderGame/LadderGame/InputView.swift:
```diff
@@ -0,0 +1,64 @@
+        guard let maxHeight:Int = Int(input) else { throw InputError.invalidType }
+        
+        writeInfoOf(maxHeight, playerNames)
+    }
```
일부 로직은 InputView 상위로 올라가야 할 것 같습니다.
InputView는 단지 입력 동작과 관련된 내용만 하고, 데이터 값을 넘겨서 흐름제어하는 것은 상위 모듈에서 구현해보세요.

godrm commented on LadderGame/LadderGame/ResultView.swift:
```diff
@@ -0,0 +1,65 @@
+            Swift.print("-----", terminator: "|")
+            return
+        }
+        Swift.print("     ", terminator: "|")
```
그냥 print() 만 쓰셔도 됩니다. 참고하세요.
내부에 만든 메소드와 헷갈리도록 이름을 만들었다면 내부 메소드 이름을 개선해보면 어떨까요? 
`print-` 대신에 다른 단어로 해도 좋겠습니다.

godrm commented on LadderGame/LadderGame/ResultView.swift:
```diff
@@ -0,0 +1,65 @@
+            return "\(player.name) "
+        default:
+            return "     "
+        }
```
이런 작업을 위해서 switch-case를 쓰는것은 데이터 개수에 따라 모두 작업해줘야 하는 상황이 됩니다. 데이터를 처리하는 로직은 개수와 상관없이 돌아가도록 일반화해보세요. 특히 switch-case (if 구문)등을 지양하세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -7,113 +7,16 @@
+do {
+    var inputView = InputView()
+    try inputView.execute()
+    let ladder = inputView.ladderGame.makeLadder()
```
위에 InputView 내용을 참고해서 이 부분을 개선해보세요.
지금은 InputView가 너무 많은 일을 하고 있습니다.

godrm reviewed (commented):
아래 피드백을 확인해보세요.

popsmile committed:
InputView를 간략화한다

변수로 가지고있던 LadderGame 구조체를 InputView와 대등하게 상위모듈로 올리기위하여, InputView내에서 안내메세지를 출력하고 입력값을 받아 문자열로 리턴하는 기능만 남겨놓는다.
구조체 내부 변수 guideMessage는 private으로 수정한다.

popsmile committed:
전달받은 입력값을 LadderGame 내부에서 제어하도록 수정한다

InputView와 동등한 단계의 모듈로 올려주기위해, 구조체 생성 시에 전달받은 값을 내부에서 제어하여 사다리배열을 리턴하도록 수정한다.

popsmile committed:
LadderGame 구조체 내부에서만 사용되는 함수는 private으로 처리한다

popsmile committed:
입력값의 유효성을 검사하는 Validator를 생성한다

참여자 이름 글자수가 5글자를 넘지않는지, 사다리 높이가 정수형인지 검사 후 발생된 에러를 main으로 전달한다

popsmile committed:
checkValidate 에러 전달시에 print될 내용을 추가한다

popsmile committed:
Validator 내부에서만 사용되는 메소드 isValid를 private 처리한다

popsmile committed:
LadderGame 외부에서도 [LadderPlayer] 가져올 수 있도록 setter만 private 처리한다

popsmile committed:
ResultView에서 내부 변수를 추가하고 이름출력 함수를 수정한다

참여자이름을 출력하기 위해 main의 Ladder Gamer에서 [LadderPlayer]만 가져오도록 수정한다.
이름 출력 함수에서 switch-case 구문을 삭제한다

popsmile committed:
main에서 InputView, LadderGame, ResultView 구조체를 사용하여 사다리게임을 실행하도록 수정한다

popsmile commented:
피드백을 참고하여 크게 아래의 내용을 작업했습니다.
1. InputView 내부 변수로 있던 LadderGame을 상위 모듈로 올려 대등한 수준으로 실행되도록 수정
2. 구조체 내부에서만 사용되는 변수 및 함수를 private으로 수정
3. 참여자 이름 출력 함수 내 switch-case 구문 삭제 및 문자열 길이에 따라 출력하도록 수정
4. 커밋 메세지를 제목과 본문으로 구분하여 작성

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,56 @@
+
+struct LadderGame {
+    private var height = 0
+    private(set) var players = [LadderPlayer]()
```
players 변수를 private(set)으로 하기 보다는
출력할 때 필요한게 참가자들 이름이니까 `[String]` 형태로 이름만 리턴해주는 메소드가 있으면 어떨까요?

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,56 @@
+            ladders.append(makeLadderRow(rowIndex: (players.count-1)))
+        }
+        return ladders
+    }
```
사다리를 만들어서 return 하는건 좋습니다만,
이렇게 지역 변수로 만들어져 있으면, makeLadder() 한 다음에 다시 사다리 정보에 접근하려면 다시 makeLadder를 호출하는 방법뿐이 없을텐데 다시 만들어지겠네요?
이렇게 호출할 때마다 1회용으로 만드는 이유는 무엇인가요?

popsmile replied:
사다리를 만들어 return하기 전에 내부에서 사다리정보에 접근해야하는 작업들을 모두 마치고, return한 이후 사다리 정보에 접근해야할 경우가 없다고 생각하여 매 게임 실행 시 새롭게 만들어지도록 하였습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -7,113 +7,19 @@
-    let ladders = makeLadders(maxHeight, numberOfPeople)
-    printLadders(ladders)
+    let playerNames = InputView(guideMessage: "참여할 사람의 이름을 입력하세요. (이름은 쉼표(,)로 구분하세요)").getInput()
+    let maxHeight = InputView(guideMessage: "최대 사다리 높이는 몇 개인가요?").getInput()
```
매개변수에 인수를 넘길 때 숫자나 지금처럼 문자열을 하드코딩하지마세요.
대신 어딘가 상수로 선언해놓고 의미있는 코드로 표현하세요.

popsmile committed:
ResultView에 전달하는 참여자이름 타입을 수정한다

ResultView에 참여자이름을 전달할 때, LadderGame내 변수를 전달하지 않고 [LadderPlayer]에서 name만 [String]으로 받아와서 전달한다

popsmile committed:
상수로 선언한 guideMessage를 매개변수로 전달하도록 한다

popsmile commented:
1. ResultView에서 [LadderPlayer] 대신 이름 문자열배열을 전달받는 방식으로 수정하였습니다.
2. 매개변수에 인수를 넘기는 경우, 하드코딩이 아닌 미리 선언해둔 상수를 사용하도록 수정하였습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -19,7 +19,7 @@ func execute() {
     
     let ladderGame = LadderGame(height: Int(maxHeight)!, names: playerNames)
-    ResultView(ladder:ladderGame.makeLadder(), players:ladderGame.players).printGame()
+    ResultView(ladder:ladderGame.makeLadder(), playerNames:ladderGame.getPlayerNames()).printGame()
```
ResultView 에서 ladder 와 playerNames 모두 printGame() 호출할 때만 필요한건가요? 
생성자에 넘기는 경우는 이후에 사용하는 메소드에서 모두 필요한 경우에 사용하면 좋고,
그렇지 않다면 printGame()을 호출할 때만 넘기는 게 좋습니다. 참고하세요.
